crates/accelerate/src/equivalence.rs

```diff
@@ -0,0 +1,786 @@
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use itertools::Itertools;
+
+use pyo3::exceptions::PyTypeError;
+use qiskit_circuit::parameter_table::ParameterUuid;
+use rustworkx_core::petgraph::csr::IndexType;
+use rustworkx_core::petgraph::stable_graph::StableDiGraph;
+use rustworkx_core::petgraph::visit::IntoEdgeReferences;
+
+use smallvec::SmallVec;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::{error::Error, fmt::Display};
+
+use exceptions::CircuitError;
+
+use ahash::RandomState;
+use indexmap::{IndexMap, IndexSet};
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PySet, PyString};
+
+use rustworkx_core::petgraph::{
+    graph::{EdgeIndex, NodeIndex},
+    visit::EdgeRef,
+};
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::imports::{ImportOnceCell, QUANTUM_CIRCUIT};
+use qiskit_circuit::operations::Param;
+use qiskit_circuit::operations::{Operation, OperationRef};
+use qiskit_circuit::packed_instruction::PackedOperation;
+
+mod exceptions {
+    use pyo3::import_exception_bound;
+    import_exception_bound! {qiskit.circuit.exceptions, CircuitError}
+}
+pub static PYDIGRAPH: ImportOnceCell = ImportOnceCell::new("rustworkx", "PyDiGraph");
+
+// Custom Structs
+
+#[pyclass(frozen, sequence, module = "qiskit._accelerate.equivalence")]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Key {
+    #[pyo3(get)]
+    pub name: String,
+    #[pyo3(get)]
+    pub num_qubits: u32,
+}
+
+#[pymethods]
+impl Key {
+    #[new]
+    #[pyo3(signature = (name, num_qubits))]
+    fn new(name: String, num_qubits: u32) -> Self {
+        Self { name, num_qubits }
+    }
+
+    fn __hash__(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        (self.name.to_string(), self.num_qubits).hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn __repr__(slf: PyRef<Self>) -> String {
+        slf.to_string()
+    }
+
+    fn __getnewargs__(slf: PyRef<Self>) -> (Bound<PyString>, u32) {
+        (
+            PyString::new_bound(slf.py(), slf.name.as_str()),
+            slf.num_qubits,
+        )
+    }
+
+    // Ord methods for Python
+    fn __lt__(&self, other: &Self) -> bool {
+        self.lt(other)
+    }
+    fn __le__(&self, other: &Self) -> bool {
+        self.le(other)
+    }
+    fn __eq__(&self, other: &Self) -> bool {
+        self.eq(other)
+    }
+    fn __ne__(&self, other: &Self) -> bool {
+        self.ne(other)
+    }
+    fn __ge__(&self, other: &Self) -> bool {
+        self.ge(other)
+    }
+    fn __gt__(&self, other: &Self) -> bool {
+        self.gt(other)
+    }
+}
+
```

Comment:
With the following, you can replace the five constructions of `Key` with `Key::from_operation` and `Key::from_gate`
```suggestion
impl Key {
    fn from_operation(operation: &PackedOperation) -> Self {
        let op_ref: OperationRef = operation.view();
        Key {
            name: op_ref.name().to_string(),
            num_qubits: op_ref.num_qubits(),
        }        
    }
    fn from_gate(gate: &GateOper) -> Self {
        Key::from_operation(&gate.operation)
    }
}

```
Or,  you might even omit `from_gate` and just do `let key = Key::from_operation(&gate.operation)`